File: AngularAndNetCoreAuth/Services/MomoService.cs
using System;
using System.Net;
using System.Net.Mime;
using System.Runtime.InteropServices;
using AngularAndNetCoreAuth.Data;
using Microsoft.IdentityModel.Protocols;
using RestSharp.Validation;
using Microsoft.Extensions.Logging;
using Newtonsoft.Json;
using System.Linq;
using RestSharp;
using AngularAndNetCoreAuth.Data;
using AngularAndNetCoreAuth.Models;
using RestSharp.Authenticators;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json.Linq;

namespace AngularAndNetCoreAuth.Services
{
    public class MomoService : IMomo
    {
        //private static string url = "http://digihub.prudentialbank.com.gh:8080/MobileMoneyPayment/api/Transaction";
        private static string url = "https://digihub.prudentialbank.com.gh/MobileMoneyPayment/api/Transaction";
        private const string clientAuth = "AD411E74-28C2-4BF9-8F9D-D5B7E22F9226";
        private const string password = "Temp123$";
        private const string username = "momoapi.user.ttu";
        private readonly ApplicationDbContext _db;
        
        private static string srmsUrl = "https://srms.ttuportal.com/api/receivePayment";
        private const string srmsAuth = "128ashbx393932";

        public MomoService(ApplicationDbContext db)
        {
            _db = db;
            
        }
        public void Authenticate()
        {
            Console.WriteLine("hello auth");
        }
        /**
         * walletType eg mtn, voda, airtiel
         * URL /DebitWallet
         */
        public int DebitWallet(string walletType, string senderName, string senderNumber, decimal amount, string transactionId,
            string remarks)
        {
            var client= new RestClient($"{url}/DebitWallet");
            var request = new RestRequest(Method.POST);
            request.AddJsonBody(new
            {
                clientId = clientAuth,
                walletType=walletType,
                senderName=senderName,
                senderNumber=senderNumber,
                amount=amount,
                transactionId=transactionId,
                transactionDate=DateTime.UtcNow,
                remarks= remarks
            
            });
            client.Authenticator = new HttpBasicAuthenticator(username, password);
            request.AddHeader("Content-type", "application/json");
            var  response =  client.Execute(request);
            var responseData = response.Content;
            HttpStatusCode statusCode = response.StatusCode;
            int numericStatusCode = (int)statusCode;
            return numericStatusCode;
            //dynamic data = JObject.Parse(responseData);
            //return Convert.ToInt16(data.status);
        }
        /**
         * URL /CreditWallet
         */
        public string CreditWallet(string walletType, string receipientName, string receipientNumber, decimal amount,
            string transactionId, string remarks)
        {
            Console.WriteLine($"{url}/CreditWallet");
            var client= new RestClient($"{url}/CreditWallet");
            var request = new RestRequest(Method.POST);
            request.AddJsonBody(new
            {
                clientId = clientAuth,
                walletType=walletType,
                receipientName=receipientName,
                senderNumber=receipientNumber,
                amount=amount,
                transactionId=transactionId,
                remarks= remarks
            
            });
            client.Authenticator = new HttpBasicAuthenticator(username, password);
            request.AddHeader("Content-type", "application/json");
            var  response =  client.Execute(request);
            var responseData = response.Content;
            return responseData;
        }
        /**
         * Get transaction status
         * URL /GetTransactionStatus
         */
        public string GetTransactionStatus(string walletType, string transactionId)
        {
            Console.WriteLine($"{url}/GetTransactionStatus");
            var client= new RestClient($"{url}/GetTransactionStatus");
            var request = new RestRequest(Method.POST);
            request.AddJsonBody(new
            {
                clientId = clientAuth,
                walletType=walletType,
                transactionId=transactionId,
                
            
            });
          
            client.Authenticator = new HttpBasicAuthenticator(username, password);
          
            request.AddHeader("Content-type", "application/json");
            var  response =  client.Execute(request);
            var responseData = response.Content;
            return responseData;
            
        }

        public string Find(Guid productId)
        {
            var productDb = _db.Product
                .FirstOrDefault(p => p.Id==productId);
 
            if (productDb == null) return "product not found";
 
             
 
            return  productDb.Name;
        }

        public string vodaAirtel(string momo)
        {
             
                        
                string[] exceptionNumber = {"026","056","027","057","020","050" };
                int index = Array.IndexOf(exceptionNumber, momo.Substring(0, 3));
                if (index > -1)
                {
                    var strPassword = Guid.NewGuid().ToString("N").Substring(0, 10);
                    return strPassword;
                }
                         
                return  Guid.NewGuid().ToString();
                         
                       
             
        }
        /**
         * send the payment to srms
         */
        public string SendPaymentToSRMS(string indexNo,decimal amount,string accountNumber,string feeType,string transactionId,DateTime transactionDate)

        {
            var client= new RestClient($"{srmsUrl}");
            var request = new RestRequest(Method.POST);
            request.AddJsonBody(new
            {
                indexno = indexNo,
                amount=amount,
                accountNumber=accountNumber,
                fee_type=feeType,
                transactionId=transactionId,
                transactionDate=transactionDate,
                auth= srmsAuth
            
            });
            //client.Authenticator = new HttpBasicAuthenticator(username, password);
            request.AddHeader("Content-type", "application/json");
            var  response =  client.Execute(request);
            var responseData = response.Content;
            var jsonResult = JsonConvert.DeserializeObject(responseData).ToString();
            var result= JsonConvert.DeserializeObject<SrmsResponse>(jsonResult);
           return result.responseCode;

        }
    }
}

File: New_Service/backend/src/main/java/com/ttu/pay/services/MomoService.java
package com.ttu.pay.services;

import com.ttu.pay.controller.StudentController;
import com.ttu.pay.model.Product;
import com.ttu.pay.payload.PBLResponse;
import com.ttu.pay.payload.StudentResponse;
import com.ttu.pay.repository.ProductRepository;
import net.minidev.json.JSONObject;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.*;
import org.springframework.web.client.RestTemplate;

import java.math.BigDecimal;
import java.nio.charset.StandardCharsets;
import java.time.LocalDateTime;
import java.util.Arrays;
import java.util.Base64;
import java.util.Optional;
import java.util.UUID;

public class MomoService implements Momo {
    private static String url = "https://digihub.prudentialbank.com.gh/MobileMoneyPayment/api/Transaction";
    private final String clientAuth = "AD411E74-28C2-4BF9-8F9D-D5B7E22F9226";
    private final String password = "Temp123$";
    private final String username = "momoapi.user.ttu";
    private static String srmsUrl = "https://srms.ttuportal.com/api/receivePayment";
    private final String srmsAuth = "128ashbx393932";
    private ProductRepository productRepository;
    private RestTemplate restTemplate;
    private static final Logger logger = LoggerFactory.getLogger(MomoService.class);

    public MomoService(ProductRepository productRepository, RestTemplate restTemplate) {
        this.productRepository = productRepository;
        this.restTemplate = restTemplate;
    }

    @Override
    public void Authenticate() {
        System.out.println("authenticate users");
    }

    @Override
    public int DebitWallet(String walletType, String senderName, String senderNumber, BigDecimal amount, String transactionId, String remarks) {

        HttpHeaders headers = new HttpHeaders();
        headers.setAccept(Arrays.asList(MediaType.APPLICATION_JSON));
        headers.setBasicAuth(username, password);
        var debitUrl = url + "/DebitWallet";
        var requestObject = new JSONObject();
        requestObject.put("clientId", clientAuth);
        requestObject.put("walletType", walletType);
        requestObject.put("senderName", senderNumber);
        requestObject.put("amount", amount);
        requestObject.put("transactionId", transactionId);
        requestObject.put("transactionDate", LocalDateTime.now());
        requestObject.put("remarks", remarks);
        HttpEntity<JSONObject> object = new HttpEntity<>(requestObject, headers);
        ResponseEntity<String> response = restTemplate.postForEntity(debitUrl, object, String.class);

        return response.getStatusCodeValue();

    }

    @Override
    public int CreditWallet(String walletType, String recipientName, String recipientNumber, BigDecimal amount, String transactionId, String remarks) {
        HttpHeaders headers = new HttpHeaders();
        headers.setAccept(Arrays.asList(MediaType.APPLICATION_JSON));
        headers.setBasicAuth(username, password);
        var debitUrl = url + "/CreditWallet";
        var requestObject = new JSONObject();
        requestObject.put("clientId", clientAuth);
        requestObject.put("walletType", walletType);
        requestObject.put("recipientName", recipientName);
        requestObject.put("senderNumber", recipientNumber);
        requestObject.put("amount", amount);
        requestObject.put("transactionId", transactionId);
        requestObject.put("remarks", remarks);
        HttpEntity<JSONObject> object = new HttpEntity<>(requestObject, headers);
        ResponseEntity<String> response = restTemplate.postForEntity(debitUrl, object, String.class);

        return response.getStatusCodeValue();

    }

    @Override
    public ResponseEntity<PBLResponse> GetTransactionStatus(String walletType, String transactionId) {
        HttpHeaders headers = new HttpHeaders();
        headers.setAccept(Arrays.asList(MediaType.APPLICATION_JSON));
        headers.setBasicAuth(username, password);
        var debitUrl = url + "/GetTransactionStatus";
        var requestObject = new JSONObject();
        requestObject.put("clientId", clientAuth);
        requestObject.put("walletType", walletType);
        requestObject.put("transactionId", transactionId);
        HttpEntity<JSONObject> object = new HttpEntity<>(requestObject, headers);
        ResponseEntity<String> response = restTemplate.postForEntity(debitUrl, object, String.class);


        return new ResponseEntity<>(PBLResponse, response.toString());
    }

    @Override
    public String FindProduct(UUID productId) {
        var product = productRepository.findById(productId).stream().filter(Product::isStatus)
                .findFirst().get();
        return product.getName();

    }

    @Override
    public String VodaAirtel(String phone) {
        return null;
    }

    @Override
    public int SendPaymentToSRMS(String indexNo, BigDecimal amount, String accountNumber, String feeType, String transactionId, LocalDateTime transactionDate) {
        HttpHeaders headers = new HttpHeaders();
        headers.setAccept(Arrays.asList(MediaType.APPLICATION_JSON));
        //headers.setBasicAuth(username, password);
        var debitUrl = srmsUrl;
        var requestObject = new JSONObject();
        requestObject.put("indexno", indexNo);
        requestObject.put("amount", amount);
        requestObject.put("accountNumber", accountNumber);
        requestObject.put("fee_type", feeType);
        requestObject.put("transactionId", transactionId);
        requestObject.put("auth", srmsAuth);
        HttpEntity<JSONObject> object = new HttpEntity<>(requestObject, headers);
        ResponseEntity<String> response = restTemplate.postForEntity(debitUrl, object, String.class);
        return response.getStatusCodeValue();
    }
}
